File: CodeDeployHook/preTrafficHook.js
// Stolen from https://aws.amazon.com/blogs/compute/implementing-safe-aws-lambda-deployments-with-aws-codedeploy/
'use strict';

const AWS = require('aws-sdk');
const codeDeploy = new AWS.CodeDeploy({ apiVersion: '2014-10-06' });
var lambda = new AWS.Lambda({httpOptions: {timeout: 30000}});

exports.handler = (event, context, callback) => {
	var debug = process.env.DEBUG === 'true';
	var isCodeDeployRun = false, forceError = false;
	if (event.DeploymentId){
		isCodeDeployRun = true;
	} else if (event.forceError == 'true'){
		forceError = true;
	}
	if (debug){console.log("Entering PreTraffic Hook!");}

	var functionToTest = process.env.FUNCTION_TO_TEST;
	if (debug){console.log("Testing new function version: " + functionToTest + ", isCodeDeployRun: " + isCodeDeployRun + ", forceError: " + forceError);}

	// Perform validation of the newly deployed Lambda version
	var lambdaResult = "Failed";
	var lambdaParams = {
		FunctionName: functionToTest,
		InvocationType: "RequestResponse",
		Payload: '{"action":"next","currentPlayer":"r","board":{"numCols":7,"numRows":6,"rows":[["r","r","y",".",".",".","."],["y","y","y",".",".",".","."],["y","y","y",".",".",".","."],["r",".","r",".",".",".","."],[".",".",".",".",".",".","."],[".",".",".",".",".",".","."]]},"column":6}'
	};
	lambda.invoke(lambdaParams, function (err, data){
		if (err){ // an error occurred
			console.error(err, err.stack);
			lambdaResult = "Failed";
		} else { // successful response
			if (debug){console.log("Result... Payload: " + data.Payload);}
			if (data.Payload == '{"gameState":"0","playerBoard":{"numCols":7,"numRows":6,"rows":[["r","r","y",".",".",".","r"],["y","y","y",".",".",".","."],["y","y","y",".",".",".","."],["r",".","r",".",".",".","."],[".",".",".",".",".",".","."],[".",".",".",".",".",".","."]]},"playerRow":0,"aiBoard":{"numCols":7,"numRows":6,"rows":[["r","r","y","y",".",".","r"],["y","y","y",".",".",".","."],["y","y","y",".",".",".","."],["r",".","r",".",".",".","."],[".",".",".",".",".",".","."],[".",".",".",".",".",".","."]]},"aiCol":3,"aiRow":0}'){
				lambdaResult = "Succeeded";
				if (debug){console.log("Validation testing succeeded!");}
			} else {
				lambdaResult = "Failed";
				console.error("Validation testing failed!");
			}

			if (isCodeDeployRun){
				// Pass AWS CodeDeploy the prepared validation test results.
				var params = {
					deploymentId: event.DeploymentId,
					lifecycleEventHookExecutionId: event.LifecycleEventHookExecutionId,
					status: lambdaResult
				};
				codeDeploy.putLifecycleEventHookExecutionStatus(params, function (err, data){
					if (err){
						// Validation failed.
						console.error('CodeDeploy Status update failed');
						console.error(err, err.stack);
						callback("CodeDeploy Status update failed");
					} else {
						// Validation succeeded.
						if (debug){console.log('CodeDeploy status updated successfully');}
						callback(null, 'CodeDeploy status updated successfully');
					}
				});
			} else if (lambdaResult == "Failed" || forceError){
				// Hand-cranked synthetic canary run and it failed
				var sns = new AWS.SNS();
				var params = {
					Message: 'Hand-cranked canary test for Connect4Trainer failed. See https://ap-southeast-2.console.aws.amazon.com/cloudwatch/home?region=ap-southeast-2#dashboards:name=Connect4',
					Subject: 'AWS:Sydney - Connect4Trainer testing call failed',
					TopicArn: process.env.SNS_TOPIC_ARN
				};
				sns.publish(params, function(err, data) {
					if (err){
						console.error('Hand-cranked canary test failed and could not send SNS topic');
						console.error(err, err.stack);
					} else {
						if (debug){console.log('Hand-cranked canary test failed and SNS topic sent!');}
					}
				});
			}
		}
	});
}

File: CodeDeployHook/preTrafficHook.py
import boto3
import logging
import os

code_deploy_client = boto3.client('codedeploy')
lambda_client = boto3.client('lambda')
sns_client = boto3.client('sns')

def handler(event, context):
	# Logger setup
	logger = logging.getLogger()
	if os.getenv('DEBUG') == 'true':
		logger.setLevel(logging.DEBUG)

	logger.debug("Entering PreTraffic Hook!")

	# Init
	code_deployment_deployment_id = None
	force_error = False
	new_function_successful = False
	if 'DeploymentId' in event:
		code_deployment_deployment_id = event['DeploymentId']
	elif 'forceError' in event and event['forceError'] == 'true':
		force_error = False
	function_to_test = os.getenv('FUNCTION_TO_TEST')

	logger.debug(f"Testing new function version: {function_to_test}, isCodeDeployRun: {code_deployment_deployment_id != None}, forceError: {force_error}")

	response = lambda_client.invoke(
		FunctionName = function_to_test,
		Payload = '{"action":"next","currentPlayer":"r","board":{"numCols":7,"numRows":6,"rows":[["r","r","y",".",".",".","."],["y","y","y",".",".",".","."],["y","y","y",".",".",".","."],["r",".","r",".",".",".","."],[".",".",".",".",".",".","."],[".",".",".",".",".",".","."]]},"column":6}'
	)

	# Check if the response from the new function is correct
	if response['StatusCode'] != 200:
		logger.error(f"New function failed. Status code: {response['StatusCode']}. Error: {response['FunctionError']}")
	else:
		payload = response['Payload'].read().decode("utf-8").strip()
		if payload == '{"gameState":"0","playerBoard":{"numCols":7,"numRows":6,"rows":[["r","r","y",".",".",".","r"],["y","y","y",".",".",".","."],["y","y","y",".",".",".","."],["r",".","r",".",".",".","."],[".",".",".",".",".",".","."],[".",".",".",".",".",".","."]]},"playerRow":0,"aiBoard":{"numCols":7,"numRows":6,"rows":[["r","r","y","y",".",".","r"],["y","y","y",".",".",".","."],["y","y","y",".",".",".","."],["r",".","r",".",".",".","."],[".",".",".",".",".",".","."],[".",".",".",".",".",".","."]]},"aiCol":3,"aiRow":0}':
			logger.info(f"New function passed successfully")
			new_function_successful = True
		else:
			logger.error(f"New function completed but response payload is wrong: {payload}")


	# Return CodeDeploy response
	if code_deployment_deployment_id is not None:
		response = code_deploy_client.put_lifecycle_event_hook_execution_status(
			deploymentId = code_deployment_deployment_id,
			lifecycleEventHookExecutionId = event['LifecycleEventHookExecutionId'],
			status = 'Succeeded' if new_function_successful else 'Failed'
		)
		if 'lifecycleEventHookExecutionId' in response:
			logger.info('Completed CodeDeploy lifecycle')
		else:
			logger.error(f"Failed to complete CodeDeploy lifecycle for deploymentId: {code_deployment_deployment_id} and lifecycleEventHookExecutionId: {event['LifecycleEventHookExecutionId']}")
	elif not new_function_successful or force_error:
		# Not doing CodeDeploy, so fire SNS if there's a failure
		logger.debug(f"Publishing failure notification to topic: {os.getenv('SNS_TOPIC_ARN')}")
		response = sns_client.publish(
			TopicArn = os.getenv('SNS_TOPIC_ARN'), # This might fail if we're not publishing to a topic in the same region
			Message = 'Hand-cranked canary test for Connect4Trainer failed. See https://ap-southeast-2.console.aws.amazon.com/cloudwatch/home?region=ap-southeast-2#dashboards:name=Connect4',
			Subject = 'Connect4Trainer testing call failed'
		)
		if 'MessageId ' not in response:
			logger.error(f"Hand-cranked canary test failed and could not send SNS topic. Response: {response}")
		else:
			logger.debug(f"Hand-cranked canary test failed and SNS topic sent. MessageId: {response['MessageId']}")
